lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -141,13 +141,20 @@ def inline_metadata(catalog, checksum_type)
 
           # TODO: Conditionally copy owner, group, and mode when we can check if permissions are set
           child_resource = Puppet::Resource.new(:file, File.join(file[:path], meta.relative_path))
+          child_resource[:ensure] = meta.ftype
           replace_metadata(child_resource, meta, true)
 
           # Copy parameters from original parent directory
           file.original_parameters.each do |param, value|
             # These should never be passed to our children
             unless [:parent, :ensure, :recurse, :recurselimit, :target, :alias, :source].include? param
-              child_resource[param] = value.to_s
+              # This logic belongs in the file type/individual parameters,
+              # e.g. param.copy(child_resource)
+              if [:before, :require, :subscribe, :notify].include? param
+                child_resource[param] = value
+              else
+                child_resource[param] = value.to_s
```

Comment:
Not sure, the [static_compiler](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/catalog/static_compiler.rb#L184) doesn't do it, and the file type passes the value to [newchild](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/file.rb#L536-L540), which ends up calling [`resource[param] = value`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type.rb#L1213-L1215). So I'd say we can drop it, but I think @HAIL9000 ran into a problem when not calling `to_s`.